File: src/botc_solver/display.py
from __future__ import annotations

from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from typing import TextIO, cast

from botc_solver.core import role_name
from botc_solver.model import World


@dataclass(frozen=True)
class ForcedRole:
    label: str
    roles: tuple[str, ...]
    missing: str = "not forced"
    include_role: bool = False


ForcedRoleSpec = object


def forced_role(
    label: object,
    roles: object | Iterable[object] | None = None,
    *,
    missing: str = "not forced",
    include_role: bool = False,
) -> ForcedRole:
    label_name = role_name(label)
    if roles is None:
        role_names = (label_name,)
    elif isinstance(roles, str) or hasattr(roles, "role_name") or hasattr(roles, "name"):
        role_names = (role_name(roles),)
    else:
        role_names = tuple(role_name(role) for role in cast(Iterable[object], roles))

    if not role_names:
        raise ValueError("At least one role is required.")

    return ForcedRole(
        label=label_name,
        roles=role_names,
        missing=missing,
        include_role=include_role,
    )


def format_solution(
    worlds: Sequence[World],
    players: Iterable[str],
    *,
    forced_roles: Iterable[ForcedRoleSpec] = (),
    forced_missing: str = "not forced",
    poison_context: str | None = None,
) -> str:
    player_order = tuple(players)
    facts = tuple(_coerce_forced_role(role, forced_missing) for role in forced_roles)

    lines = [f"{len(worlds)} satisfying world(s)"]
    for index, world in enumerate(worlds, start=1):
        lines.append("")
        lines.append(f"World {index}")
        lines.extend(_format_world_lines(world, player_order, poison_context=poison_context))

    if facts:
        lines.append("")
        lines.append("Forced facts")
        for fact in facts:
            lines.append(f"  {_format_forced_role(worlds, fact)}")

    return "\n".join(lines)


def print_solution(
    worlds: Sequence[World],
    players: Iterable[str],
    *,
    forced_roles: Iterable[ForcedRoleSpec] = (),
    forced_missing: str = "not forced",
    poison_context: str | None = None,
    file: TextIO | None = None,
) -> None:
    print(
        format_solution(
            worlds,
            players,
            forced_roles=forced_roles,
            forced_missing=forced_missing,
            poison_context=poison_context,
        ),
        file=file,
    )


def _coerce_forced_role(role: ForcedRoleSpec, forced_missing: str) -> ForcedRole:
    if isinstance(role, ForcedRole):
        return role
    return forced_role(role, missing=forced_missing)


def _format_world_lines(
    world: World,
    players: Sequence[str],
    *,
    poison_context: str | None,
) -> list[str]:
    lines: list[str] = []
    for player in players:
        actual = world.actual_role(player)
        apparent = world.apparent.get(player)
        poison_suffix = " poisoned" if world.is_poisoned(player, poison_context) else ""
        if apparent and apparent != actual:
            lines.append(f"  {player}: {actual} (appears as {apparent}){poison_suffix}")
        else:
            lines.append(f"  {player}: {actual}{poison_suffix}")
    return lines


def _format_forced_role(worlds: Sequence[World], role: ForcedRole) -> str:
    assignment = _forced_assignment(worlds, role.roles)
    if assignment is None:
        return f"{role.label}: {role.missing}"

    holder, actual_role = assignment
    if role.include_role:
        return f"{role.label}: {holder} ({actual_role})"
    return f"{role.label}: {holder}"


def _forced_assignment(
    worlds: Sequence[World],
    roles: Sequence[str],
) -> tuple[str, str] | None:
    assignments = {_world_assignment(world, roles) for world in worlds}
    if len(assignments) != 1:
        return None

    return next(iter(assignments))


def _world_assignment(world: World, roles: Sequence[str]) -> tuple[str, str] | None:
    assignments = [
        (holder, role)
        for role in roles
        for holder in [world.holder(role)]
        if holder is not None
    ]
    if len(assignments) == 1:
        return assignments[0]
    return None

File: src/display.ts
import { type RoleRef, roleName } from "./core";
import type { World } from "./model";

export interface ForcedRole {
  readonly label: string;
  readonly roles: readonly string[];
  readonly missing: string;
  readonly includeRole: boolean;
}

export type ForcedRoleSpec = RoleRef | ForcedRole;

export function forcedRole(
  label: RoleRef,
  roles?: RoleRef | readonly RoleRef[],
  options: { readonly missing?: string; readonly includeRole?: boolean } = {},
): ForcedRole {
  const labelName = roleName(label);
  let roleNames: string[];
  if (roles === undefined) roleNames = [labelName];
  else if (Array.isArray(roles)) roleNames = (roles as readonly RoleRef[]).map((role) => roleName(role));
  else roleNames = [roleName(roles as RoleRef)];
  if (roleNames.length === 0) throw new Error("At least one role is required.");
  return {
    label: labelName,
    roles: roleNames,
    missing: options.missing ?? "not forced",
    includeRole: options.includeRole ?? false,
  };
}

export function formatSolution(
  worlds: readonly World[],
  players: readonly string[],
  options: {
    readonly forcedRoles?: readonly ForcedRoleSpec[];
    readonly forcedMissing?: string;
    readonly poisonContext?: string;
  } = {},
): string {
  const facts = (options.forcedRoles ?? []).map((role) =>
    coerceForcedRole(role, options.forcedMissing ?? "not forced"),
  );
  const lines = [`${worlds.length} satisfying world(s)`];
  worlds.forEach((world, index) => {
    lines.push("", `World ${index + 1}`, ...formatWorldLines(world, players, options.poisonContext));
  });
  if (facts.length > 0) {
    lines.push("", "Forced facts");
    for (const fact of facts) lines.push(`  ${formatForcedRole(worlds, fact)}`);
  }
  return lines.join("\n");
}

export function printSolution(
  worlds: readonly World[],
  players: readonly string[],
  options: {
    readonly forcedRoles?: readonly ForcedRoleSpec[];
    readonly forcedMissing?: string;
    readonly poisonContext?: string;
  } = {},
): void {
  console.log(formatSolution(worlds, players, options));
}

function coerceForcedRole(role: ForcedRoleSpec, forcedMissing: string): ForcedRole {
  if (typeof role === "object" && role !== null && "roles" in role && "label" in role) return role as ForcedRole;
  return forcedRole(role as RoleRef, undefined, { missing: forcedMissing });
}

function formatWorldLines(world: World, players: readonly string[], poisonContext?: string): string[] {
  return players.map((player) => {
    const actual = world.actualRole(player);
    const apparent = world.apparent.get(player);
    const poisonSuffix = world.isPoisoned(player, poisonContext) ? " poisoned" : "";
    return apparent !== undefined && apparent !== actual
      ? `  ${player}: ${actual} (appears as ${apparent})${poisonSuffix}`
      : `  ${player}: ${actual}${poisonSuffix}`;
  });
}

function formatForcedRole(worlds: readonly World[], role: ForcedRole): string {
  const assignment = forcedAssignment(worlds, role);
  if (assignment === undefined) return `${role.label}: ${role.missing}`;
  const [holder, actualRole] = assignment;
  return role.includeRole ? `${role.label}: ${holder} (${actualRole})` : `${role.label}: ${holder}`;
}

function forcedAssignment(worlds: readonly World[], role: ForcedRole): readonly [string, string] | undefined {
  const assignments = new Map<string, readonly [string, string] | undefined>();
  for (const world of worlds) {
    const assignment = worldAssignment(world, role.roles);
    assignments.set(role.includeRole ? JSON.stringify(assignment) : JSON.stringify(assignment?.[0]), assignment);
  }
  return assignments.size === 1 ? [...assignments.values()][0] : undefined;
}

function worldAssignment(world: World, roles: readonly string[]): readonly [string, string] | undefined {
  const assignments = roles.flatMap((role) => {
    const holder = world.holder(role);
    return holder === undefined ? [] : [[holder, role] as const];
  });
  return assignments.length === 1 ? assignments[0] : undefined;
}
